fix(snownet,relay): re-use channels to peers in cooldown period

thomaseizinger opened the pull request:
For efficiency reasons, TURN's data channels don't have any authentication or integrity metadata. Instead, the operate using a short 2-byte channel number to identify the target peer of the data.

To avoid abuse, channel bindings are at most valid for 10 minutes before they need to be refreshed. In case they expire, there is a 5 minute cooldown period, before the same channel number can be bound to a different peer and before the same peer can be bound to a different channel.

We had a similar issue in the past (#5613) where channels got rebound early. Whilst that was fixed and is no longer happening, a case that we didn't consider is what happens if we want to bind a channel to a peer that still has a channel bound but is currently cooling down (i.e. in the 5 minute period after its expiry).

In that case, `snownet` would wrongly assume that there is no channel to this peer and try to bind a new one. That would get rejected by the relay with a bad request.

To fix this, we simply need to check whether we still have a channel to this peer and if yes, return the same channel number. On the relay, we need to ensure that we consider a channel as `bound` again when it is being refreshed.

We ensure that this doesn't regress in two ways:

- We add a unit-test for the `ChannelBindings` struct
- We modify the `Idle` transition to idle for 6 instead of 5 minutes. This ensures that a combination of 2 idle transitions puts the channel bindings into the 10-15 minute time window where rebinding the peer to a different channel fails.

Related: #6265.

thomaseizinger committed:
Re-use channels that are cooling down

thomaseizinger committed:
Add changelog entries

jamilbk commented on website/src/components/Changelog/GUI.tsx:
```diff
@@ -13,12 +13,17 @@ export default function GUI({ title }: { title: string }) {
+          <ChangeItem pull="6276">
+            Fixes a bug where relayed connections failed to establish after an idle period.
+          </ChangeItem>
+        </ul>
```
```suggestion
          <ChangeItem pull="6276">
            Fixes a bug where relayed connections failed to establish after an idle period.
          </ChangeItem>
```

jamilbk replied:
repeated the `<ul>` here I think by accident

thomaseizinger replied:
Ah yes, my HTML-fu is a bit dated.

jamilbk commented on website/src/components/Changelog/Android.tsx:
```diff
@@ -15,10 +15,17 @@ export default function Android() {
+          <ChangeItem pull="6276">
+            Fixes a bug where relayed connections failed to establish after an idle period.
+          </ChangeItem>
+        </ul>
```
```suggestion
          <ChangeItem pull="6181">
            Improves reliability of DNS resolution of non-resources.
          </ChangeItem>
          <ChangeItem pull="6276">
            Fixes a bug where relayed connections failed to establish after an idle period.
          </ChangeItem>
        </ul>
```

`ChangeItem`s are `<li>`s

thomaseizinger committed:
Only use one list

thomaseizinger committed:
fix(snownet): invalidate allocation on channel binding error (#6277)

When we receive an `AllocationMismatch` error response from the relay,
it means that our local state is toast and needs to be invalidated.

- If we attempted to allocate, the corrective action is to delete the
active allocation.
- If we attempted to refresh or bind a channel, the corrective action is
to make a new allocation.

In the case of a channel binding, we also re-schedule the target peer to
be rebound to ensure upper layers don't need to retry that. For example,
if this happens during a connection setup, we still want to eventually
succeed in binding the channel to ensure STUN messages as part of ICE
can pass over it without having to first run into an ICE timeout and
retry the entire connection.

In certain network configurations, we observed that the NAT between
connlib and the relay may have fairly short session timers. Currently,
allocations have a lifetime of 10 minutes and are refreshed every 5
minutes. If there is no other traffic from connlib during those 5
minutes, the NAT session might get cut and attempting to use the
allocation to e.g. bind a channel doesn't work because the relay doesn't
recognise the 3-tuple.

We deem these situations quite rare. Instead of keeping the NAT session
alive with additional traffic, we instead implement this corrective
action here which transparently creates a new allocation using our new
3-tuple.

Resolves: #6265.
